index.rst

```diff
@@ -13,17 +13,151 @@
+Similarly, ``dax_apdb`` does not provide a way to specify its code version, it has to be extended as well to include that information (independently for SQL and Cassandra implementations).
+
+
+Version recording
+-----------------
+
+The database needs to record the versions with which it was created (or later upgraded).
+One common approach for this is to define a separate metadata table that can record various additional information about the database itself.
+This table can be a simple key-value storage wit two string columns, indexed by a key value, this approach is used for ``daf_butler`` schema management tools.
+Both versions used to create database schema will be recorded in the metadata table, one possible example choosing names for keys could be::
+
+    key                           | value
+   -------------------------------+----------
+    version:sdm_schemas/apdb.yaml | 1.2.0
+    version:dax_apdb/ApdbSql      | 2.0.1
+
+Addition of this metadata table to the existing databases would be a schema change in itself, which could probably be managed by the migration tools.
+
+
+Schema migration tools
+----------------------
+
+Schema upgrades for existing databases will require a migration tool which will know about all existing versions of the schema and corresponding scripts to migrate from one version to later versions.
+Different backends (SQL and Cassandra) may share some of the migration tool functionality but will likely have a different set of migration scripts.
+For SQL-based implementation it is natural to use `Alembic`_ as it solves many of the migration issues.
+`Alembic`_ cannot be used directly with Cassandra and it is unlikely that it can be adapted for use with Cassandra with a reasonable effort, more likely Cassandra backend will need a separate tooling.
+
+Many ideas from `daf_butler_migrate`_ can be reused for implementing APDB migration tool, including command line interface and organization of the Alembic migration scripts.
+Cassandra-specific tooling can be added to the same command-line tool to provide a uniform interface, if possible.
+
+
+Implementation path
+===================
+
+If above model looks reasonable then implementation plan for adding schema versions to APDB could look like:
+
+
+- Extend ``felis`` to support versions in schema definitions as a schema-level ``version`` key.
+  These versions can be any strings, ``felis`` is not going to interpret them.
+  We could use semantic version numbers for APDB schemas or provide more explicit specification of compatibility in ``felis`` (e.g. ``compatible_versions`` key).
+
+- Add a starting version number to ``apdb.yaml``, ``1.0.0`` may be a good start, leaving ``0.x.y`` space for pre-metadata history.
+
+- Add a starting version numbers to ``ApdbSql`` and ``ApdbCassandra`` classes.
+  Extend these classes with option of reading stored version numbers from metadata table, if that exists, and check compatibility.
```

Comment:
I'm assuming that compatibility check would raise an exception on connecting if the databases differ in MAJOR number, warn on connecting and raise on write if they differ by MINOR and either warn or do nothing if they differ by PATCH?

Reply:
If we use semantic versioning, then patch indeed should not matter, there is no need to issue warnings for patch differences. Difference in minor version _may_ mean read compatibility, that is up to us to agree what the meaning is. I'd probably avoid warnings when things are compatible, it can cause warning flooding. Major version changes should be considered incompatible.

Semantic versioning is not the only option (and is not frequently used for database schemas), for `felis` schemas we could use just a sequential numbering and specify their read/write compatibility explicitly.

Reply:
Ok, please note how those different cases would be handled (do nothing, vs. warning vs. raise) in a couple of sentences in the text here.